src/hotspot/share/gc/g1/g1CollectionSetCandidates.hpp

```diff
@@ -29,72 +29,199 @@
+
+  uint length() const { return (uint)_candidates.length(); }
+
+  void verify() PRODUCT_RETURN;
+
+  // Comparison function to order regions in decreasing GC efficiency order. This
+  // will cause regions with a lot of live objects and large remembered sets to end
+  // up at the end of the list.
+  static int compare(CandidateInfo* ci1, CandidateInfo* ci2);
+
+  G1CollectionCandidateListIterator begin() {
+    return G1CollectionCandidateListIterator(this, 0);
+  }
 
-  HeapRegion* at(uint idx) const {
-    HeapRegion* res = nullptr;
-    if (idx < _num_regions) {
-      res = _regions[idx];
-      assert(res != nullptr, "Unexpected null HeapRegion at index %u", idx);
-    }
-    return res;
+  G1CollectionCandidateListIterator end() {
+    return G1CollectionCandidateListIterator(this, length());
   }
+};
+
+// Iterator for G1CollectionSetCandidates.
+class G1CollectionSetCandidatesIterator : public StackObj {
+  G1CollectionSetCandidates* _which;
+  uint _marking_position;
+
+public:
+  G1CollectionSetCandidatesIterator(G1CollectionSetCandidates* which, uint marking_position);
+
+  G1CollectionSetCandidatesIterator& operator++();
+  HeapRegion* operator*();
+
+  bool operator==(const G1CollectionSetCandidatesIterator& rhs);
+  bool operator!=(const G1CollectionSetCandidatesIterator& rhs);
+};
 
-  // Remove num_regions from the front of the collection set candidate list.
-  void remove(uint num_regions);
-  // Remove num_remove regions from the back of the collection set candidate list.
-  void remove_from_end(uint num_remove, size_t wasted);
+// Tracks all collection set candidates, i.e. regions that could/should be evacuated soon.
+//
+// These candidate regions are tracked in a list of regions, sorted by decreasing
+// "gc efficiency".
+//
+// Currently there is only one type of such regions:
+//
+// * marking_regions: the set of regions selected by concurrent marking to be
+//                    evacuated to keep overall heap occupancy stable.
+//                    They are guaranteed to be evacuated and cleared out during
+//                    the mixed phase.
+//
+class G1CollectionSetCandidates : public CHeapObj<mtGC> {
+  friend class G1CollectionSetCandidatesIterator;
+
+  enum class CandidateOrigin : uint8_t {
+    Invalid,
+    Marking,                   // This region has been determined as candidate by concurrent marking.
+    Verify                     // Special value for verification.
+  };
+
+  G1CollectionCandidateList _marking_regions;
+
+  CandidateOrigin* _contains_map;
+  uint _max_regions;
 
-  // Iterate over all remaining collection set candidate regions.
-  void iterate(HeapRegionClosure* cl);
-  // Iterate over all remaining collection set candidate regions from the end
-  // to the beginning of the set.
-  void iterate_backwards(HeapRegionClosure* cl);
+  // The number of regions from the last merge of candidates from the marking.
+  uint _last_marking_candidates_length;
```

Comment:
Looking at how it is used, I wonder if we can record `min_old_cset_length`, which is what is actually needed.

Reply:
I would like to defer this suggestion (which is good) as this is an improvement of the existing code which I would like to follow here for this change.